File: leetcode_91_medium/91.decode-ways.cpp
/*
 * @lc app=leetcode id=91 lang=cpp
 *
 * [91] Decode Ways
 */

// @lc code=start
#include<string>
using namespace std;
class Solution {
public:
    int numDecodings(string s) {
        return search(s);
    }
    int search(string other) {
        if (other.size()==1) {
            if (int(other[0])-48 > 0) {
                return 1;
            } else {
                return 0;
            }
        }
        if (other.size() >=2) {
            int l,r =0;
            auto leftValue = int(other[0]) - 48;
            if (leftValue > 0) {
                auto left = other.substr(1,other.size()-1);
                l = search(left)+1;
            }
            auto rightValue = (int(other[0]) - 48)*10+(int(other[1]) - 48);
            if (rightValue>=10 && rightValue <= 26) {
                auto right = other.substr(2,other.size()-2);
                r = search(right)+1;
            }
            return max(l,r);
        }
        return 0;
    }
};
// @lc code=end

int main() {
    auto s = Solution();
    printf("%d", s.numDecodings("06"));
}

File: leetcode_91_medium/91.decode-ways.py
#
# @lc app=leetcode id=91 lang=python3
#
# [91] Decode Ways
#

# @lc code=start
class Solution:
    def numDecodings(self, s: str) -> int:
        if len(s) == 0 :
            return 0
        opt = [0 for i in range(0,len(s)+1)]
        opt[0] = 1
        if int(s[0]) > 0:
            opt[1] = 1
        for i in range(1, len(s)):
            v1 = 0
            v2 = 0
            if int(s[i]) > 0:
                v1 = opt[i]
            if i-1 >= 0 and (int(s[i-1:i+1]) >=10 and int(s[i-1:i+1]) <=26) :
                v2 = opt[i-1]
            opt[i+1] = v1+v2
        return opt[len(s)]
# @lc code=end

if __name__ == "__main__":
    ss = Solution()
    